packages/monitor-v2/src/monitor-og/oSnapAutomation.ts

```diff
@@ -274,16 +286,82 @@ const filterVerifiedProposals = async (
+  const executedAssertionIds = (
+    await Promise.all(
+      Array.from(new Set(proposals.map((proposal) => proposal.address))).map(async (ogAddress) => {
+        const og = await getOgByAddress(params, ogAddress);
+        const executedProposals = await runQueryFilter<ProposalExecutedEvent>(og, og.filters.ProposalExecuted(), {
+          start: 0,
+          end: params.blockRange.end,
+        });
+        return executedProposals.map((executedProposal) => executedProposal.args.assertionId);
+      })
+    )
+  ).flat();
+
+  // Filter out all proposals that have been executed based on matching assertionId.
+  return proposals.filter((proposal) => !executedAssertionIds.includes(proposal.args.assertionId));
+};
+
+// Filter function to check if challenge period has passed for a proposal.
+const hasChallengePeriodEnded = (proposal: TransactionsProposedEvent, timestamp: number): boolean => {
+  return timestamp >= proposal.args.challengeWindowEnds.toNumber();
+};
+
+// Filters supported proposal events and adds their parameters to the result.
+const getSupportedProposals = async (
+  proposals: TransactionsProposedEvent[],
+  params: MonitoringParams
+): Promise<SupportedProposal[]> => {
+  // Get OOv3 for checking if assertion's bond is supported.
+  const oo = await getOo(params);
+
+  // Keep only proposals whose rules are parsable and bond is supported based on its assertionId.
+  const supportedProposals = (
+    await Promise.all(
+      proposals.map(async (event) => {
+        const parsedRules = parseRules(event.args.rules);
+        const { currency, bond } = await oo.getAssertion(event.args.assertionId);
+        const isSupported = parsedRules !== null && isBondSupported(currency, bond.toString(), params.supportedBonds);
+        return isSupported ? { event, parameters: { parsedRules, currency, bond: bond.toString() } } : null;
+      })
+    )
+  ).filter((proposal) => proposal !== null) as SupportedProposal[];
+
+  return supportedProposals;
+};
+
+// Filter proposals that did not pass verification and also retain verification result for logging.
+const getDisputableProposals = async (
+  proposals: SupportedProposal[],
+  params: MonitoringParams
+): Promise<DisputableProposal[]> => {
+  // TODO: We should separately handle IPFS and Graphql server errors. We don't want to submit disputes immediately just
+  // because IPFS gateway or Snapshot backend is down.
+  return (
+    await Promise.all(
+      proposals.map(async (proposal) => {
+        const verificationResult = await verifyProposal(proposal.event, params);
+        return !verificationResult.verified ? { ...proposal, verificationResult } : null;
+      })
+    )
+  ).filter((proposal) => proposal !== null) as DisputableProposal[];
+};
+
 const approveBond = async (
   provider: Provider,
   signer: Signer,
   currency: string,
   bond: string,
   spender: string
 ): Promise<void> => {
-  // If bond is 0, no need to approve.
-  if (bond === "0") return;
```

Comment:
We don't expect 0 bonds in production, but this allows covering edge cases in testing where bond value is changed just before submitting proposal.